### ISSUE TITLE
Angular 2.0 should have a (documented) a11y story.

### ISSUE THREAD

**Issue by matanlurey** [angular/angular#1878]
In Angular 1.0 Dart, it was necessary to add all kind of wild hacks to get this support, e.g.

```dart
class MaterialRadioButton {
  @override
  void attach() {
    _disposer.addDisposable(node.onClick.listen((_) => changeSelection()));
    _disposer.addDisposable(node.onKeyDown.listen(handleKeyDown));

    // outline for keyboard navigation
    _disposer.addDisposable(node.onKeyUp.listen((KeyboardEvent keyEvent) {
      if (keyEvent.keyCode == KeyCode.TAB) {
        _ngElement.removeClass("mouse");
      }
    }));
    _disposer.addDisposable(node.onMouseDown.listen((_) {
      _ngElement.addClass("mouse");
    }));

    // add subs for propper group control keyboard
    if (_group != null) {
      _group.registerRadioButton(this);
      _disposer.addDisposable(node.onFocus.listen((_) {
        _group.onFocus(this);
      }));
      _disposer.addDisposable(node.onBlur.listen((_) {
        _group.onBlur(this);
      }));
    }
  }

  set tabIndex(String tabIndex) {
    if (tabIndex == null) {
      _ngElement.removeAttribute("tabIndex");
    } else {
      _ngElement.setAttribute("tabIndex", tabIndex);
    }
  }

void handleKeyDown(KeyboardEvent keyEvent) {
    if (keyEvent.keyCode == KeyCode.SPACE) {
      changeSelection();
    }

    if (_group != null) {
      if (keyEvent.keyCode == KeyCode.LEFT || keyEvent.keyCode == KeyCode.UP) {
        _group.selectPrevious(this);
        keyEvent.preventDefault();
        _ngElement.removeClass("mouse");
      } else if (keyEvent.keyCode == KeyCode.RIGHT ||
          keyEvent.keyCode == KeyCode.DOWN) {
        _group.selectNext(this);
        keyEvent.preventDefault();
        _ngElement.removeClass("mouse");
      }
    }
  }
}
```

We would like to be able to follow the [Google accessibility guidelines](https://sites.google.com/a/google.com/accessibility/design-guide/standard-keyboard-conventions). Ideally, Angular should offer examples and user guides on how to do this in a non-hacky way.

@jelbourn, who may already have answers for some of this.

Please refer to internal bug for more information about use cases.

**Comment by jelbourn** [comment]
I don't quote follow what you mean by "wild hacks". 
(I've never used Angular dart, just AngularJS) 

In the AngularJS 1.X world, the framework was mostly agnostic to a11y (before ngAria)- it didn't do anything to make it easier, but it also didn't make it any harder. As far as I see it now, Angular 2 is in about the same boat. Did you have something specific in mind for what you'd want angular to do for you?

If you want some examples, you can check out the current radio buttons in `modules/angular_material` to see what I've done. I see that kind of thing as just part of building the component.

FYI @marcysutton, a11y expert for the existing `angular/material`.

**Comment by marcysutton** [comment]
One of the biggest issues with Angular 1 was lack of discouragement for binding `ng-click` to `div` and other non-semantic elements. Although you could just as easily bind a click event to a div with native JavaScript, Angular can encourage best practices (like only ng-clicking interactive elements like buttons) and avoid the widespread accessibility issues previously encountered. 

I agree that a documented accessibility story would be great! The best thing would be to just include accessible examples throughout the framework, and lead by example, rather than siloing accessibility into a single document. The discussions around extending native elements and supporting roles/attributes in Angular 2 show accessibility is being considered (and a module like ngAria shouldn't be necessary), but a lot of it is up to documentation.

**Comment by matanlurey** [comment]
"Crazy hacks" = anti-patterns (injecting element and doing changes yourself)

**Comment by jelbourn** [comment]
Is manually doing stuff with the element considered an anti-pattern in angular dart? I thinking is that Angular makes a lot of stuff easier for the application engineer, but there are always going to be times when you need to do something more complex that requires you to get into the underlying platform.

I think Angular 2 will make some things easier, such as being able to set up listeners for _any_ event. 

I don't know if there are enough universal behaviors for accessible components such that you could leverage the framework to do anything of significant value.

**Comment by matanlurey** [comment]
Manually editing the element makes code  *mostly* non-portable in Angular 2. This is the prime issue.

**Comment by jelbourn** [comment]
AFAIK, the "portability" for Angular 2 is meant for the framework itself. I don't think the goal is to be able to build components that can run on any arbitrary platform.

**Comment by naomiblack** [comment]
Internally, this is also b/21118157

**Comment by marcysutton** [comment]
In order to say "Angular 2 has ARIA support" and "the ngAria module will no longer be needed in Angular 2", the framework should handle the same use cases as ngAria. Some of these won't directly apply with the new framework architecture, but I wanted to make sure they were captured.

Accessibility use cases for Angular include:

* Handling `aria-disabled` when disabling custom controls
* Handling `aria-*` value attributes with ngModel
* Discouraging click events on non-semantic `div` elements (could be done through documentation)
* Labels for form controls (also a documentation issue)

**Comment by jakemoves** [comment]
Chiming in to support that the docs should strongly discourage binding click events to non-semantically clickable elements (like `div`).

**Comment by naomiblack** [comment]
Assigning to @marcysutton who is working on a cookbook chapter for the dev guides on a11y that covers this.

**Comment by jakemoves** [comment]
Awesome! @marcysutton if it's useful, I'd be happy to be a test reader for what you put together. (If it's not useful, please ignore this comment :D )

**Comment by marcysutton** [comment]
Thanks @jakemoves, when it's time for testing we'll be sure to let you know!

**Comment by hmig** [comment]
@marcysutton:  Something I’ve seen that folks doing is `<a href=“#”>` and adding click handling …. plz add that to the list of what not to do. I was lazy during a project and left fixing those to buttons until the end AND it became a nightmare. Lesson learned, but the devs should have made them `<buttons>` to begin with.

**Comment by mhevery** [comment]
This is obsolete, and no longer relevant / actionable. To keep our issues clean, we are aggressively closing them.

**Comment by guikubivan** [comment]
So where is this guide? Also @hmig, how did you fix the `<a href="#">`, did you make them buttons?

**Comment by omouse** [comment]
I'm also reading through the guide and I see "accessibility" mentioned in the features of Angular 2, but not in the docs anywhere else.
